### ISSUE TITLE
Liquidsoap utilizes 100% CPU and stops streaming

### ISSUE THREAD

**Issue by binslash** [savonet/liquidsoap#485]
I'm using Liquidsoap to stream to an Icecast server on localhost. After a while (this could be a handful of songs, or over an hour) Liquidsoap will begin to decode the next song, but never finish.

When this occurs the cpu utilization jumps from 5-10% to 100% and Icecast kills the stream (I assume because it isn't receiving any data). At this point, trying to stop it via SIGINT causes it to try to wait until the threads finish. They never do, so I have to SIGSTOP and kill -9 it.

There doesn't seem to be any correlation between tracks that cause this. Tracks that it will lock-up on, will play perfectly when I restart the program.

I've tried using both a playlist and directories to no avail.

I use MPD to stream to Icecast with no issues, I believe this to be something related to Liquidsoap.

Here's some info on my setup:

Debian 9.1
Liquidsoap 1.3.2 (compiled via ocaml, development libraries are from the Debian repro)
Icecast 2.4.2 (from Debian repo)

Liquidsoap script:
[main.liq.txt](https://github.com/savonet/liquidsoap/files/1340716/main.liq.txt)

Liquidsoap log:
[main.log](https://github.com/savonet/liquidsoap/files/1340719/main.log)

**Comment by toots** [comment]
@binslash do you have a specific time in the log that I should look at?

**Comment by binslash** [comment]
@toots At line 1293:
`2017/09/27 16:12:08 [beets_rename:4] Remaining: 29.6s, queued: 30.0s, adding: 30.0s (RID 2)`
Is where liquidsoap jumps to 100% utilization. The next line (1294) is where I sent a SIGINT. Liquidsoap hangs with "Waiting for threads to terminate..."

**Comment by toots** [comment]
Ok, thanks. Your script looks relatively minimal and you're using the latest version. Sounds like we should be able to find this issue. 

Have you ever used `gdb`? It would be great to have a stack trace of the issue. Basically, I think the following should work:
- Start your liquidsoap instance
- Wait for the issue to show up
- Attach a gdb inspector to it: `gdb -p <LIQ PID>`

At this point, `gdb` should freeze the program and you can then run:
```
thread apply all bt
```

That should give us an idea what's being executed inside liquidsoap when the issue occurs.

Thanks!

**Comment by davehayes** [comment]
Here's another datapoint, admittedly from FreeBSD. (If you aren't familiar with FreeBSD, just ask. I am.)

It is fairly easy for me to rig up a test case where a `kill -9` is required.

[liqreport.txt](https://github.com/savonet/liquidsoap/files/1376890/liqreport.txt)

**Comment by toots** [comment]
@davehayes First, thanks for taking the time! Then: shouldn't this apply to #483 instead?

**Comment by davehayes** [comment]
@toots Maybe. I strongly suspect they are two separate issues, but I haven't verified this.

**Comment by toots** [comment]
Yeah me too, that's why I wondered if you meant to post this to the other issue :-)

I might have something for you to test shortly, tho!

**Comment by davehayes** [comment]
Ok I've verified (through my naive first attempt at writing a VLC decoder) that this bug is not related to decoding WMA files. I see the same behavior when this decoder failed:

```
2017/10/11 13:33:39 [decoder:4] Trying method "VLC" for "random.wma"...
/usr/local/bin/vlc -vvv --no-repeat --no-loop -I dummy random.wma --sout '#transcode{acodec=s16l,channels=2,samplerate=44100}:file{mux=wav,dst=/tmp/L3RlcmFyYWlkL211c2ljL0xpYnJhcnkvbW9yZW11c2ljL1BhdCBNZXRoZW55L1BhdCBNZXRoZW55IC0gIE9uIEhlciBXYXkud21h.wav}' vlc://quit 2>/dev/null
2017/10/11 13:33:39 [clock:4] Main phase starts.
2017/10/11 13:33:39 [allfmt2(dot)m3u:5] Queue is empty!
2017/10/11 13:33:39 [allfmt2(dot)m3u:5] Failed to prepare track: no file.
2017/10/11 13:33:39 [allfmt2(dot)m3u:5] Queue is empty!
2017/10/11 13:33:39 [allfmt2(dot)m3u:5] Failed to prepare track: no file.
```

**Comment by toots** [comment]
Hey @davehayes ! Let's try and keep our conversation on #485. I'm still not sure that @binslash's issue is that same and I'd like to wait for his backtrace to resume here. I'll comment more on your issue.

**Comment by binslash** [comment]
I've a log of the gdb trace below.

[liquidsoap_gdb.log](https://github.com/savonet/liquidsoap/files/1382851/liquidsoap_gdb.log)

**Comment by toots** [comment]
Thanks so much for this @binslash! I'm still a little in the dark but it helps a lot. Would you be able to test on a specific git branch? I'll give you instruction on how to build it..

**Comment by toots** [comment]
Ok, I've done a big cleanup on the `flac` code. Not sure if it fixes your bug but it's worth trying. Here's how you can do it:
```
git clone --recursive https://github.com/savonet/ocaml-flac.git
cd ocaml-flac
opam pin add flac .
```
That should recompile `flac` and, incidentally, `liquidsoap` with the latest code.

**Comment by binslash** [comment]
@toots I tried using the git code, but it won't compile. I've attached a log of the output.

[ocaml-flac.log](https://github.com/savonet/liquidsoap/files/1387403/ocaml-flac.log)

**Comment by toots** [comment]
You might need some additional dev tools. In Debian or Ubuntu these would be `autoconf` and `automake`..

**Comment by binslash** [comment]
I was able to recompile, but the same issue is happening.

**Comment by toots** [comment]
hi @binslash,

I've had another pass at cleaning up the `ocaml-flac` code. Any chance you could try again with the latest `git` code? If you're using `opam` pin method, this should work:
```
% cd /path/to/ocaml-flac
% git pull
% opam reinstall flac
```

Thanks!

**Comment by binslash** [comment]
@toots 

Nope, same issue.

Are you sure this is a flac issue? I've had the same issue happen with mp3 files as well.

Thanks

**Comment by toots** [comment]
That's what the stack trace shows. The flac decoder might still be tested against mp3 files. 

Just to be sure, though, would you mind running another `gdb` stack trace against the latest code? That would help a lot..

Thanks!

**Comment by binslash** [comment]
I've attached a stack trace.

[nov14_liquidsoap_gdb.log](https://github.com/savonet/liquidsoap/files/1472487/nov14_liquidsoap_gdb.log)

**Comment by toots** [comment]
Thanks. I'll let you know if I have more ideas..

**Comment by toots** [comment]
Also, anything special about your hardware? What kind of server is it?

**Comment by binslash** [comment]
It's a Dell R710 with 96 GB of ram and dual Xeon L5520 running VMware ESXi 6.0.0, 3073146

This is transparent to the VM, which I've allocated 1 GB of ram, a thick provisioned 8 GB hard disk, 100% usable CPU with a 1 socket, 1 core virtual CPU.

**Comment by toots** [comment]
Ok, thanks. I'm gonna try to reproduce locally.

**Comment by toots** [comment]
Hi,

Still not many ideas on my part here but I'll try to have another look at it shortly.

**Comment by binslash** [comment]
I tried a Fedora VM, and Raspbian running on a Raspberry Pi, but the same issue happened.
I've been updating the Debian VM and testing it out for any changes every few weeks. About a week ago, it seemed to start working. I was able to run it without issue all weekend.

I checked apt and opam logs, but I didn't see anything related to FLAC, or even audio, that would account for the change...

Well, it's working now. Should I close this, or do you still want to look in to it?

**Comment by toots** [comment]
Hmm interesting. Any chance in the OCaml compiler?

**Comment by toots** [comment]
I think this was fixed by commit e6159a6. Closing.
